tests/unit/test_id_cli.py

```diff
@@ -0,0 +1,466 @@
+                yield tokens.duplicate(workflow=id_)
+
+        monkeypatch.setattr(
+            'cylc.flow.id_cli._expand_workflow_tokens_impl',
+            _expand_workflow_tokens_impl,
+        )
+
+    _patch_expand_workflow_tokens(['xxx'])
+    return _patch_expand_workflow_tokens
+
+
+@pytest.mark.parametrize(
+    'ids_in,ids_out,multi_mode',
+    [
+        # multi mode should be True if multiple workflows are defined
+        (['a//'], ['a'], False),
+        (['a//', 'b//'], ['a', 'b'], True),
+        # or if pattern matching is used, irrespective of the number of matches
+        (['*//'], ['xxx'], True),
+    ]
+)
+async def test_parse_ids_multi_mode(
+    patch_expand_workflow_tokens,
+    ids_in,
+    ids_out,
+    multi_mode,
+):
+    """It should glob for workflows.
+
+    Note:
+        More advanced tests for this in the integration tests.
+
+    """
+    workflows, _multi_mode = await parse_ids_async(
+        *ids_in,
+        constraint='workflows',
+        match_workflows=True,
+    )
+    assert list(workflows) == ids_out
+    assert _multi_mode == multi_mode
+
+
+@pytest.fixture
+def src_dir(tmp_path):
+    """A src dir containing a workflow called "a"."""
+    cwd_before = Path.cwd()
+    src_dir = (tmp_path / 'a')
+    src_dir.mkdir()
+    src_file = src_dir / 'flow.cylc'
+    src_file.touch()
+    os.chdir(tmp_path)
+    yield src_dir
+    os.chdir(cwd_before)
+
+
+def test_parse_src_path(src_dir):
+    """It should locate src dirs."""
+    # valid absolute path
+    workflow_id, src_path, src_file_path = _parse_src_path(
+        str(src_dir.resolve())
+    )
+    assert workflow_id == 'a'
+    assert src_path == src_dir
+    assert src_file_path == src_dir / 'flow.cylc'
+
+    # broken absolute path
+    with pytest.raises(UserInputError):
+        workflow_id, src_path, src_file_path = _parse_src_path(
+            str(src_dir.resolve()) + 'xyz'
+        )
+
+    # valid relative path
+    workflow_id, src_path, src_file_path = _parse_src_path('./a')
+    assert workflow_id == 'a'
+    assert src_path == src_dir
+    assert src_file_path == src_dir / 'flow.cylc'
+
+    # broken relative path
+    with pytest.raises(WorkflowFilesError):
+        _parse_src_path('.')
+
+    # relative '.' (invalid)
+    with pytest.raises(WorkflowFilesError) as exc_ctx:
+        workflow_id, src_path, src_file_path = _parse_src_path('.')
+    assert 'No flow.cylc or suite.rc in .' in str(exc_ctx.value)
+
+    # move into the src dir
+    os.chdir(src_dir)
+
+    # relative '.' (valid)
+    workflow_id, src_path, src_file_path = _parse_src_path('.')
+    assert workflow_id == 'a'
+    assert src_path == src_dir
+    assert src_file_path == src_dir / 'flow.cylc'
+
+    # relative './<flow-file>'
+    workflow_id, src_path, src_file_path = _parse_src_path('./flow.cylc')
+    assert workflow_id == 'a'
+    assert src_path == src_dir
+    assert src_file_path == src_dir / 'flow.cylc'
+
+
+async def test_parse_ids_src_path(src_dir):
```

Comment:
Async tests are getting skipped on GH Actions: https://github.com/cylc/cylc-flow/runs/4741547845?check_suite_focus=true#step:11:2594
```
/opt/hostedtoolcache/Python/3.9.9/x64/lib/python3.9/site-packages/_pytest/python.py:172:
  PytestUnhandledCoroutineWarning: async def functions are not natively supported
  and have been skipped.
  You need to install a suitable plugin for your async framework, for example:
    - anyio
    - pytest-asyncio
    - pytest-tornasync
    - pytest-trio
    - pytest-twisted
    warnings.warn(PytestUnhandledCoroutineWarning(msg.format(nodeid)))
```

Reply:
Hmmm.

I used to put an explicit `@pytest.mark.asyncio` onto these functions which got them running but I recently found that they ran even without that marker.

I've gotta strange feeling that it's because I've got `pytest-tornasync` installed in my environment (due to UIS development), from poking at the Jupyter Server tests I noticed that `pytest-tornasync` seems to enable itself by default so I guess I'm probably running unmarked async tests through `pytest-tornasync` (which is `asyncio` under the hood)?

Dunno, I guess I will go through and add explicit `pytest.mark.asyncio` lines?

Reply:
There might be a `pytest-asyncio` config...

Reply:
...Yes but it won't be available until the next release...